src/KeyEntityFormEnhancer.php

```diff
@@ -498,10 +509,20 @@ private function createSuggestion(\Exception $exception, KeyInterface $key): Mar
       // valid organization name and username provided with an invalid password
       // the MGMT server returns HTTP 500 with an error instead of HTTP 401.
       if ($exception->getCode() === 401 || ($exception->getCode() === 500 && $exception->getEdgeErrorCode() === 'usersandroles.SsoInternalServerError')) {
-        $suggestion = $this->t('@fail_text The given username (%username) or password is incorrect.', [
-          '@fail_text' => $fail_text,
-          '%username' => $key_type->getUsername($key),
-        ]);
+
+        // If on public cloud (using the default endpoint), the username should
+        // be an email.
+        if ($key_type->getEndpointType($key) == 'default' && !$this->emailValidator->isValid($key_type->getUsername($key))) {
```

Comment:
Does this change compatible with the environment variables API key storage? Because maybe the new entity_type field's definition is missing from [here](https://github.com/apigee/apigee-edge-drupal/blob/8.x-1.x/src/Plugin/KeyType/ApigeeAuthKeyType.php#L42-L75) and thanks for that the environment credentials storage [does not expose](https://github.com/apigee/apigee-edge-drupal/blob/8.x-1.x/src/Plugin/KeyProvider/EnvironmentVariablesKeyProvider.php#L115) an environment variable for this field. IOW, what happens if I set the value of the APIGEE_EDGE_USERNAME environment variable to an invalid email and the APIGEE_EDGE_ENDPOINT to a private org (without being able the set the endpoint type so it remains "default"). If I click on "Test connection, does it generate a suggestion that the username should be an email address?

If this is real bug then probably the same issue occurs for the `authorization_server_type` introduced in #254.

Reply:
I've updated the `EdgeKeyTypeBase::getEndpointType()` method so that if the `endpoint_type` property isn't defined, it still returns the correct type, that way we can avoid adding more environment variables.

https://github.com/apigee/apigee-edge-drupal/blob/ac7ba5cf219ac0807685af4f4acc4b652b444e80/src/Plugin/EdgeKeyTypeBase.php#L73-L77

`authorization_server_type` isn't an issue, because it isn't used outside of the `ApigeeAuthKeyInput` class/form.